Enable encrypted raw sending to pools with greater ashift

gamanakis opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Closes #13067.

### Description
<!--- Describe your changes in detail -->
Raw sending from pool1/encrypted with ashift=9 to pool2/encrypted with
ashift=12 results to failure when mounting pool2/encrypted (Input/Output
error). Notably, the opposite, raw sending from a larger ashift to a lower one
does not fail.

This happens because zio_compress_write() falsely checks only
ZIO_FLAG_RAW_COMPRESS and not ZIO_FLAG_RAW_ENCRYPT which is also set in
encrypted raw send streams. In this case it rounds up the psize and if
not equal to the zio->io_size it modifies the block by zeroing out
the extra bytes. Because this happens in a SA attr. registration object
(type=46), the decryption fails upon mounting the filesystem, and zpool
status falsely reports an error.

Fix this by checking both ZIO_FLAG_RAW_COMPRESS and ZIO_FLAG_RAW_ENCRYPT
before deciding whether to zero-pad a block.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->
A test in ZTS has been added.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

jumbi77 commented on module/zfs/zio.c:
```diff
@@ -1773,7 +1773,13 @@ zio_write_compress(zio_t *zio)
+	} else if (zio->io_flags & ZIO_FLAG_RAW_COMPRESS &&
+	    !(zio->io_flags & ZIO_FLAG_RAW_ENCRYPT)) {
+		/*
+		 * If we are raw receiving an ecnrypted dataset we should not
```
nit: s/ecnrypted/encrypted/

gamanakis replied:
Thanks, fixed!

gamanakis commented:
[e351356](https://github.com/openzfs/zfs/pull/13074/commits/e351356e966190f1feed78c49842b9da246bad02): Fix a typo.

gamanakis commented:
This is ready for review, it's an one line fix in zio.c.

behlendorf commented on tests/zfs-tests/tests/functional/rsend/send_raw_ashift.ksh:
```diff
@@ -0,0 +1,192 @@
+{
+	rm -f $BACKDIR/fs@*
+	datasetexists pool9 && log_must zpool destroy pool9
+	datasetexists pool12 && log_must zpool destroy pool12
```
```suggestion
	poolexists pool12 && destroy_pool pool12
```

behlendorf commented on tests/zfs-tests/tests/functional/rsend/send_raw_ashift.ksh:
```diff
@@ -0,0 +1,192 @@
+	datasetexists pool9 && log_must zpool destroy pool9
+	datasetexists pool12 && log_must zpool destroy pool12
+	rm -f $TESTDIR/vdev_a
+	rm -f $TESTDIR/vdev_b
```
```suggestion
	log_must rm -f $TESTDIR/vdev_a $TESTDIR/vdev_b
```

behlendorf commented on tests/zfs-tests/tests/functional/rsend/send_raw_ashift.ksh:
```diff
@@ -0,0 +1,192 @@
+{
+	log_must zfs set xattr=sa pool9/$1
+	log_must zfs set dnodesize=legacy pool9/$1
+	log_must zfs set recordsize=128k pool9/$1
```
It like like you don't need to enable compression on the dataset to get `ZIO_FLAG_RAW_COMPRESS` set, but it still might be a good idea.  To improve the test coverage you randomly set some of these dataset properties.

behlendorf reviewed (commented):
Makes sense, thanks for digging in to this.

gamanakis commented:
Thank you for **your** efforts @behlendorf.
[6491793](https://github.com/openzfs/zfs/commit/6491793c920ce062c26ce01a08d26701950a7bb3): Rebased to master, addressed suggestions.

gamanakis commented:
[7b29f73](https://github.com/openzfs/zfs/commit/7b29f7364ccb00d3f1f44ab3e4f735f8a8663398): Rebased to master.

isopix commented:
It sounds like it fixed only half of the problem(or description is not correct), because it sounds like raw encrypted send to lower ashift will still corrupt data (but maybe it's only description)?

gamanakis commented:
@isopix actually raw sending to lower ashift was never broken (confirmed with testing). Only sending to greater ashift, which this PR fixed.
